interface/main/main_screen.php

```diff
@@ -90,87 +90,136 @@ function generate_html_end()
-                array($_SESSION['authId'])
+                "UPDATE users_secure SET login_work_area = ? WHERE id = ?",
+                array($requests, $userid)
             );
-        } catch (u2flib_server\Error $e) {
-            // Authentication failed so we will build the U2F form again.
-            $form_response = '';
-            $errormsg = xl('Authentication error') . ": " . $e->getMessage();
+            generate_html_start(xl('U2F Key Verification'));
+            echo "<p>\n";
+            echo xlt('Insert your key into a USB port and click the Authenticate button below.');
+            echo " " . xlt('Then press the flashing button on your key within 1 minute.') . "</p>\n";
+            echo "<table><tr><td>\n";
+            echo "<input type='button' value='" . xla('Authenticate') . "' onclick='doAuth()' />\n";
+            echo "<input type='hidden' name='form_requests' value='" . attr($requests) . "' />\n";
+            echo "<input type='hidden' name='form_response' value='' />\n";
+            echo "</td></tr></table>\n";
+            if ($errormsg) {
+                echo "<p style='color:red;font-weight:bold'>" . text($errormsg) . "</p>\n";
+            }
+            exit(generate_html_end());
         }
-    }
-    if (!$form_response) {
-        // There is no key data yet or authentication failed, so we need to solicit it.
-        $requests = json_encode($u2f->getAuthenticateData($registrations));
-        // Persist the challenge also in the database because the browser is untrusted.
-        sqlStatement(
-            "UPDATE users_secure SET login_work_area = ? WHERE id = ?",
-            array($requests, $userid)
-        );
-        generate_html_start(xl('U2F Key Verification'));
-        echo "<p>\n";
-        echo xlt('Insert your key into a USB port and click the Authenticate button below.');
-        echo " " . xlt('Then press the flashing button on your key within 1 minute.') . "</p>\n";
-        echo "<table><tr><td>\n";
-        echo "<input type='button' value='" . xla('Authenticate') . "' onclick='doAuth()' />\n";
-        echo "<input type='hidden' name='form_requests' value='" . attr($requests) . "' />\n";
-        echo "<input type='hidden' name='form_response' value='' />\n";
-        echo "</td></tr></table>\n";
-        if ($errormsg) {
-            echo "<p style='color:red;font-weight:bold'>" . text($errormsg) . "</p>\n";
+    } elseif ($registrations["method"] == "TOTP") {
+        if ($row1['method'] == 'TOTP') {
+            $registrations['secret'] = $row1['var1'];
+        }
+        $errormsg = false;
+        if ($form_response) {
+
+            require_once("../../library/classes/Totp.class.php");
+            $passwordResults = privQuery(
+                "SELECT password FROM users_secure WHERE username = ?",
+                array($_POST["authUser"])
+            );
+            $password = $passwordResults["password"];
```

Comment:
I know this is anal but would not place the password in this variable if don't need to. How about doing the following below instead:
```
$googleAuth = new Totp($passwordResults["password"], $registrations['secret']);
```

Reply:
Okay, I'll change this.